### PR TITLE
Allow symlinks to library dictionary path

### PR DESCRIPTION
**Pull request by kssenii**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow symlinks for library dictionaty path.

### CODE REVIEW

File: src/Common/filesystemHelpers.cpp [hunk 1]
```diff
@@ -122,6 +123,17 @@ bool pathStartsWith(const std::filesystem::path & path, const std::filesystem::p
     return path_starts_with_prefix_path;
 }
 
+bool symlinkStartsWith(const std::filesystem::path & path, const std::filesystem::path & prefix_path)
+{
+    auto absolute_path = std::filesystem::absolute(path);
```
Comment (kssenii) [inline src/Common/filesystemHelpers.cpp:128]: The point is not to use fs::canonical or fs::weakly_canonical

**Review by tavplubix — CHANGES_REQUESTED** [review]
Now it does not check anything. It will return true for `path = "/var/lib/clickhouse/user_files/../../../../../../../etc/passwd"` and `prefix_path = "/var/lib/clickhouse/user_files"`

**Comment by kssenii** [comment]
Then we cannot allow symlinks for library dictionary path? Because the only way to fix this is to use fs::canonical, but for symlink it will return canonized path which is not inside dictionaries path...

**Comment by tavplubix** [comment]
> Then we cannot allow symlinks for library dictionary path? Because the only way to fix this is to use fs::canonical (or weakly_canonical), but for symlink it will return canonized path which is not inside dictionaries path...

It used to work earlier somehow, so we can.

**Comment by kssenii** [comment]
There as a check `if (!startsWith(path, dictionaries_lib_path))` in previous versions which was not also correct.

**Comment by kssenii** [comment]
Oh, looks like `fs::lexically_normal` will also fix this.

File: src/Common/filesystemHelpers.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Common/filesystemHelpers.cpp:127]: I feel like I need a comment to:
* function `symlinkStartsWith` itself - about how it works with symlinks
* probably, every call for `std::filesystem`, cause I need to go to documentation to understand what exactly it does; and also why it is needed here

Comment (KochetovNicolai) [reply src/Common/filesystemHelpers.cpp:127]: BTW, can't we just canonise both path and then compare?

Comment (kssenii) [reply src/Common/filesystemHelpers.cpp:127]: > can't we just canonise both path and then compare?

This is exactly what was done and why it did not work. Applying fs::canonical or fs::weakly_canonical to a symlink results in a canonical path -- where it links to, but what we need is a path of the symlink itself.

**Review by KochetovNicolai — APPROVED** [review]
LGTM
@tavplubix what do you think?

**Review by tavplubix — APPROVED** [review]

**Pull request merged**